### PR TITLE
tests: add libgpgme to valgrind suppressions

### PR DESCRIPTION
**Pull request by petermax2**
relates to #2341 .

## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy.
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

**Comment by mpranj** [comment]
Thank you so much for still investigating this. I hope this was the missing suppression. :crossed_fingers:

**Comment by petermax2** [comment]
> Thank you so much for still investigating this. I hope this was the missing suppression. crossed_fingers

I highly doubt it. I will also try to initialize the library only once from within the unit test.

**Comment by petermax2** [comment]
I will re-run the Jenkins tests for about 3 times and see if I can trigger any of the errors mentioned in #2341 .

**Comment by petermax2** [comment]
Now one of the crypto tests failed. I'll try to set `GNUPGHOME` from within each unit test (gpgme, crypto, fcrypt) as this could maybe (?!) be concurrency issue?!

**Comment by mpranj** [comment]
Quite possible. I had not considered that yet.

**Comment by petermax2** [comment]
If I understand `tests.c` correctly, every single test case gets its own `HOME` directory, is that correct?

**Comment by mpranj** [comment]
> If I understand tests.c correctly, every single test case gets its own HOME directory, is that correct?

Afaik the temporary HOME is initialized only once for a set of plugin tests. If no other trickery is done this means that there is one HOME per `testmod_xyz`, but a different set of tests `testmod_abc` would have their own home.

**Comment by petermax2** [comment]
My current idea is to exclude the memory checks of crypto, fcrypt, and gpgme from regular test runs and create separate build taks where the memory checker runs exclusively for crypto, fcrypt and gpgme.

**Comment by petermax2** [comment]
Instead of bringing stability the crypto plugin now seems to fail consistently when being mem-checked in a separated build task. This phenomenon fascinates me.

**Comment by petermax2** [comment]
### Status update

With separation into several build task enabled, we now have a consistently occurring memory leak when running `testmod_crypto` under Fedora 31:

    ==2829== Command: /usr/bin/gpg2 --batch --trust-model always -d
    ==2829== Parent PID: 2817
    ==2829== 

and still some spontaneous occurring errors under Debian:

    CRYPTO       TESTS
    ==================

    crypto Results: 65 Tests done — 0 errors.
    unexpected flag: 3
    /home/jenkins/workspace/libelektra_PR-3425/tests/cframework/tests.c:523: error in clean_temp_home: Could not delete TMPHOME via nftw

This error is quite notable. I think that in this case a socket might still be in use by the `gpg-agent`. It seems like the agent has not shut down yet and thus released its socket.

**Comment by mpranj** [comment]
> This error is quite notable. I think that in this case a socket might still be in use by the gpg-agent. It seems like the agent has not shut down yet and thus released its socket.

I think this last error has not (yet) randomly occurred before.

On a positive note, the builds are green right now. I will trigger a rebuild to check how consistent this is.

**Comment by petermax2** [comment]
Random memory leak by `gpg` in Debian Sid

    ==3744== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
    ==3744== Command: /usr/bin/gpg2 --batch --trust-model always -d
    ==3744== Parent PID: 3730

@mpranj I'm out of ideas. Do we leave the memory check for crypto, fcrypt and gpgme disabled? What are your thoughts?

**Comment by mpranj** [comment]
> @mpranj I'm out of ideas. Do we leave the memory check for crypto, fcrypt and gpgme disabled? What are your thoughts?

Don't worry too much. It seems we can not find the underlying problem(s) in a reasonable amount of time.

I do not consider the leak as critical as having green builds. Thus I would disable the crypto memchecks completely on the problematic instances, such that other PRs and master builds are not broken.

I suggest opening a PR where all crypto plugins are enabled and just keeping an eye on what can be done. It's definitely not highest-priority to fix this right now and we don't even know whether problem is just upstream.

**Comment by petermax2** [comment]
Jenkins build please

**Comment by petermax2** [comment]
I agree with @mpranj . I'll try some more mem-check runs with separate Debian build tasks but if any further error occurs I will simply disable the affected memcheck runs and open a new PR with crypto memchecks enabled.

**Comment by petermax2** [comment]
Latest build was green, I restarted the Jenkins build.

**Comment by mpranj** [comment]
Builds are green right now. 👍 Good job! Ping me when it's ready.

**Comment by petermax2** [comment]
You were faster than me. I will start one last Jenkins build over night. Let's continue tomorrow.

**Comment by petermax2** [comment]
This time crypto did not fail but `build-elektra-web-base` did:

    The command '/bin/sh -c cmake .. -DTOOLS="kdb;web" && make -j ${PARALLEL} && make install' returned a non-zero code: 2

One more try.

**Comment by mpranj** [comment]
Sometimes the go packages fail to load:

```
go: github.com/google/uuid@v1.1.1: Get https://proxy.golang.org/github.com/google/uuid/@v/v1.1.1.mod: net/http: TLS handshake timeout
```

**Comment by petermax2** [comment]
> Sometimes the go packages fail to load

our CI is not the most stable one :laughing:

File: scripts/jenkins/Jenkinsfile [hunk 1]
```diff
@@ -477,7 +478,7 @@ def generateMainBuildStages() {
       CMAKE_FLAGS_DEBUG +
       CMAKE_FLAGS_BUILD_FULL +
       CMAKE_FLAGS_BUILD_STATIC + [
-      'PLUGINS': 'ALL;-crypto',
+      'PLUGINS': 'ALL',
```
Comment (mpranj) [inline scripts/jenkins/Jenkinsfile:481]: No need to change this now but just fyi: CMAKE_FLAGS_BUILD_ALL already contains `'PLUGINS': 'ALL'` among other things, so you only need to provide this line if you want to override the default.

I'll clean it up in a different PR, priority is to merge this sooner and get more stable builds.

**Comment by petermax2** [comment]
The last build was green again. The builds should be more stable now. I think we can merge.

**Comment by mpranj** [comment]
Thank you so much! Let’s hope this brings more stability to the builds.

**Pull request merged**